### ISSUE TITLE
Test to see if Function or Mixin Exists

### ISSUE THREAD

**Issue by Snugug** [sass/sass#561]
While I know that the general consensus whenever I want more program language-ie stuff in Sass is that it's not a programming language, with the complexity of Sass systems that currently exist and being developed, I truly feel as if we need to add some more advanced program language esque stuff into Sass to make everything work well together. To this extent, I'd really like to see `function-exists()` and `mixin-exists()` functions, or simply an `exists()` function, that will return a boolean `true` or `false` depending on whether or not that function or mixin is indeed available to use when that function gets called.

There are, IMO, countless usecases for this, but the one very specifically that I'm running in to right now is that Compass's `$base-font-size` only accepts single values, where as the grid system Singularity's `$base-font-size` accepts multiple values to allow for font scaling and has a companion function to find which base font size to use. If I'm writing a function to return an `rem` value, I need to know the proper base font size. If Singularity's base font size finder function exists, I can safely use that function to find the base font size regardless of if it's a single value or multiple values, but if I assume one or the other, the results for the function won't always be accurate, which is a problem. The solution to this is to have a check in place to see if that exists.

**Comment by robwierzbowski** [comment]
+1 from me. I don't think there's going to be any shortage of use cases for an `exists()` function, just the decision whether this type of more programmy construct belongs in Sass. Right now I'm writing code that uses a variable as a flag to tell a wrapper mixin to use either a user-set mixin or a fallback mixin -- with `exists()` I could just test for the user-set mixin and do away with the variable entirely. 

If not as part of Sass itself, this would be a really useful extension.

**Comment by boram** [comment]
+1 from me as well. I have a ruby gem that needs a mixin named clearfix. Since its defined in Compass and Bourbon, it'd be nice to check if its already been defined by those frameworks.

**Comment by edwardoriordan** [comment]
> To this extent, I'd really like to see function-exists() and mixin-exists() functions, or simply an exists() function, that will return a boolean true or false depending on whether or not that function or mixin is indeed available to use when that function gets called.

I created an extension that lets you do this -> https://github.com/edwardoriordan/sass-utilities. It requires sass 3.3.x alpha but that is for variable-exists type functionality (and variable interpolation that is also provides). The code to test for functions and mixins, however, would work fine in sass 3.2.x.

It works by simply querying the environment object to see if the named function or mixin exists. Like

``` ruby
def mixin_exists(mixin_name)
    if(environment.mixin(mixin_name.value))
      Sass::Script::Bool.new(true)
    else
      Sass::Script::Bool.new(false)
    end
  end
```

If maintainers what to add exists type functionality and the above is a sensible solution am happy to submit a pull request.

**Comment by robwierzbowski** [comment]
I started working on this earlier with a more Ruby knowledgeable friend. Here's [our notes](https://gist.github.com/robwierzbowski/4400237); we haven't had time to do much in the last few months.

**Comment by chriseppstein** [comment]
:thumbsup: also `variable-exists($varname, $global)`. I think I prefer a function per-type instead of a generic exists function. Thoughts, @nex3?

**Comment by nex3** [comment]
What's the value of `variable-exists` as opposed to just `$varname == null`? Also, what does the `$global` flag do?

**Comment by robwierzbowski** [comment]
Sass throws an error if you try to do anything with a non-existant variable. The only way to test if a var exists is by setting `$var = somevalue !default` and then testing for the value you've set, which is a bit hacky.

**Comment by nex3** [comment]
In that case, sure, `variable-exists` seems fine. Still not sure what's up with `$global`, though.

**Comment by chriseppstein** [comment]
@nex3 the $global flag, when set to true would ask whether the global variable exists (ignores local scopes).

**Comment by nex3** [comment]
When would you ever want that to be false? Local variables are in the current lexical scope, so you can always tell visually if they exist.

**Comment by chriseppstein** [comment]
because if you've set a variable of the same name locally, but need to check if the global exists, this will return true when it'd be false otherwise. Granted, this is an edge case. But if we only check for globals the behavior of `$foo: 1px; variable-exists($foo) == false` would be strange to a user.

**Comment by chriseppstein** [comment]
(Note: this is assuming we change the scoping behavior as planned)

**Comment by nex3** [comment]
Could we just name the function `global-variable-exists`?

**Comment by chriseppstein** [comment]
Naming is everything :) :+1:

**Comment by robwierzbowski** [comment]
What are the planned changes for scope? Looked around the issue queue, didn't see anything. Javascript-like?

[edit] Ah, probably https://github.com/nex3/sass/issues/136.

**Comment by nex3** [comment]
I'd describe it as python-like. When setting a variable anywhere but the top level, it will default to being lexically scoped to the enclosing block, even if it has the same name as a global variable. In order to set a global variable in a block, you'll need to write `$variable: value !global`.

**Comment by chriseppstein** [comment]
This is very useful in conjunction with #626.

**Comment by ericdfields** [comment]
Just wanna give my thanks for working on this guys. Any updates? Anything to test out?

Thanks again.

**Comment by chriseppstein** [comment]
Merged to master as 4e515ab6a46507013fcbe426a5ecfa061b4ae571
